lib/HLSL/DxilValidation.cpp

```diff
@@ -3958,111 +3958,105 @@ static void ValidateBitcode(ValidationContext &ValCtx) {
 static void ValidateWaveSize(ValidationContext &ValCtx,
                              const hlsl::ShaderModel *SM, Module *pModule) {
   // Don't do this validation if the shader is non-compute
-  if (SM->IsCS() || SM->IsLib()) {
-
-    NamedMDNode *EPs = pModule->getNamedMetadata("dx.entryPoints");
-    if (EPs != nullptr) {
-      for (unsigned i = 0, end = EPs->getNumOperands(); i < end; ++i) {
-        MDTuple *EPNodeRef = dyn_cast<MDTuple>(EPs->getOperand(i));
-        if (EPNodeRef->getNumOperands() < 5) {
-          ValCtx.EmitMetaError(EPNodeRef, ValidationRule::MetaWellFormed);
+  if (!(SM->IsCS() || SM->IsLib()))
+    return;
+
+  NamedMDNode *EPs = pModule->getNamedMetadata("dx.entryPoints");
+  if (!EPs)
+    return;
+
+  for (unsigned i = 0, end = EPs->getNumOperands(); i < end; ++i) {
+    MDTuple *EPNodeRef = dyn_cast<MDTuple>(EPs->getOperand(i));
+    if (EPNodeRef->getNumOperands() < 5) {
+      ValCtx.EmitMetaError(EPNodeRef, ValidationRule::MetaWellFormed);
+      return;
+    }
+    // get access to the digit that represents the metadata number that
+    // would store entry properties
+    const llvm::MDOperand &mOp =
+        EPNodeRef->getOperand(EPNodeRef->getNumOperands() - 1);
+    // the final operand to the entry points tuple should be a tuple.
+    if (mOp == nullptr || (mOp.get())->getMetadataID() != Metadata::MDTupleKind)
+      continue;
+
+    // get access to the node that stores entry properties
+    MDTuple *EPropNode = dyn_cast<MDTuple>(
+        EPNodeRef->getOperand(EPNodeRef->getNumOperands() - 1));
+    // find any incompatible tags inside the entry properties
+    // increment j by 2 to only analyze tags, not values
+    bool foundTag = false;
+    for (unsigned j = 0, end2 = EPropNode->getNumOperands(); j < end2; j += 2) {
+      const MDOperand &propertyTagOp = EPropNode->getOperand(j);
+      // note, we are only looking for tags, which will be a constant
+      // integer
+      bool invalidPropertyTag =
+          propertyTagOp == nullptr || (propertyTagOp.get())->getMetadataID() !=
+                                          Metadata::ConstantAsMetadataKind;
+      DXASSERT(!invalidPropertyTag,
+               "tag operand should be a constant integer.");
+      if (invalidPropertyTag) {
```

Comment:
I think we're misusing `assert` here still. We're asserting that a condition is never true, but then we have code that handles the possibility of it being true.

If this condition really can't happen we should have an `assert` and we shouldn't need the continue. If it can happen, we should do something to report the error.